In [1]:
import os

# Clone the repository if it hasn't been cloned already
repo_name = 'oss-transparency-python'
if not os.path.exists(repo_name):
    !git clone https://github.com/nikhilgg/oss-transparency-python.git

# Navigate to the specific directory
# This path is relative to the cloned repository root
directory_path = os.path.join(repo_name, 'outputs', 'tables')

# List the contents of the directory
if os.path.exists(directory_path):
    print(f"Contents of {directory_path}:")
    !ls -lh {directory_path}
else:
    print(f"Directory not found: {directory_path}")

Contents of oss-transparency-python/outputs/tables:
total 29M
-rw-r--r-- 1 root root  41K Mar  7 03:27 clean_repo_level_busfactor.csv
-rw-r--r-- 1 root root  11M Mar  7 03:27 clean_repo_month_panel.csv
-rw-r--r-- 1 root root  57K Mar  7 03:27 clean_vuln_discovery_rate.csv
-rw-r--r-- 1 root root 358K Mar  7 03:27 clean_vuln_survival.csv
-rw-r--r-- 1 root root 2.2K Mar  7 03:27 correlation_matrix_panel_active.csv
-rw-r--r-- 1 root root 4.1K Mar  7 03:27 correlation_matrix_repo_level.csv
-rw-r--r-- 1 root root  42K Mar  7 03:27 dataset_repo_level_busfactor.csv
-rw-r--r-- 1 root root  11M Mar  7 03:27 dataset_repo_month_panel.csv
-rw-r--r-- 1 root root  57K Mar  7 03:27 dataset_vuln_discovery_rate.csv
-rw-r--r-- 1 root root  19K Mar  7 03:27 dataset_vuln_quarterly.csv
-rw-r--r-- 1 root root 355K Mar  7 03:27 dataset_vuln_survival.csv
-rw-r--r-- 1 root root 834K Mar  7 03:27 github_bug_issues.csv
-rw-r--r-- 1 root root 682K Mar  7 03:27 github_contributors.csv
-rw-r--r-- 1 root root 2.5M Ma

In [2]:
import pandas as pd
import os

# Ensure directory_path is defined from previous steps
# If you ran the previous cell, directory_path should be available.
# For safety, re-define if not present (e.g., if you skipped steps)

# directory_path was defined in cell 1d14df37 as:
# directory_path = os.path.join(repo_name, 'outputs', 'tables')

# Load dataset_repo_month_panel.csv
'''
file_path_1 = os.path.join(directory_path, 'clean_repo_month_panel.csv')
if os.path.exists(file_path_1):
    dataset_repo_month_panel = pd.read_csv(file_path_1)
    print(f"Loaded {os.path.basename(file_path_1)}. First 5 rows:")
    display(dataset_repo_month_panel.head())
else:
    print(f"File not found: {file_path_1}")


# Load dataset_repo_level_busfactor.csv
file_path_2 = os.path.join(directory_path, 'dataset_repo_level_busfactor.csv')
if os.path.exists(file_path_2):
    dataset_repo_level_busfactor = pd.read_csv(file_path_2)
    print(f"Loaded {os.path.basename(file_path_2)}. First 5 rows:")
    display(dataset_repo_level_busfactor.head())
else:
    print(f"File not found: {file_path_2}")
'''
# Load dataset_vuln_discovery_rate.csv
file_path_3 = os.path.join(directory_path, 'clean_vuln_discovery_rate.csv')
if os.path.exists(file_path_3):
    dataset_vuln_discovery_rate = pd.read_csv(file_path_3)
    print(f"Loaded {os.path.basename(file_path_3)}. First 5 rows:")
    display(dataset_vuln_discovery_rate.head())
else:
    print(f"File not found: {file_path_3}")
'''
# Load dataset_vuln_survival.csv
file_path_4 = os.path.join(directory_path, 'dataset_vuln_survival.csv')
if os.path.exists(file_path_4):
    dataset_vuln_survival = pd.read_csv(file_path_4)
    print(f"Loaded {os.path.basename(file_path_4)}. First 5 rows:")
    display(dataset_vuln_survival.head())
else:
    print(f"File not found: {file_path_4}")
'''


Loaded clean_vuln_discovery_rate.csv. First 5 rows:


,repo_full_name,quarter,vuln_count,vuln_critical,vuln_high,vuln_medium,vuln_low,vuln_severe,stars,forks,log_stars,transparency_index,governance_index,governance_artifact_score,scorecard_score,external_contributor_ratio,bus_factor_proxy_k50,gini_contrib,top1_share,contributor_count
0,Azure/azure-sdk-for-python,2022-Q3,1,0,0,1,0,0,5472.0,3226.0,8.607582,0.964687,0.686,0.8,6.1,0.92,8.0,0.637572,0.159007,98.0
1,Azure/azure-sdk-for-python,2024-Q2,1,0,0,1,0,0,5472.0,3226.0,8.607582,0.964687,0.686,0.8,6.1,0.92,8.0,0.637572,0.159007,98.0
2,Delgan/loguru,2022-Q1,2,0,1,0,0,1,23570.0,767.0,10.067772,0.885047,0.476,0.2,6.6,0.95,1.0,0.909701,0.824121,72.0
3,Kludex/starlette,2023-Q1,1,1,0,0,0,1,11902.0,1113.0,9.384546,1.000000,0.080,0.2,NaN,1.00,2.0,0.803834,0.338039,100.0
4,Kludex/starlette,2023-Q2,3,0,1,0,0,1,11902.0,1113.0,9.384546,1.000000,0.080,0.2,NaN,1.00,2.0,0.803834,0.338039,100.0


'\n# Load dataset_vuln_survival.csv\nfile_path_4 = os.path.join(directory_path, \'dataset_vuln_survival.csv\')\nif os.path.exists(file_path_4):\n    dataset_vuln_survival = pd.read_csv(file_path_4)\n    print(f"Loaded {os.path.basename(file_path_4)}. First 5 rows:")\n    display(dataset_vuln_survival.head())\nelse:\n    print(f"File not found: {file_path_4}")\n'

In [3]:
'''
print("Data types of dataset_repo_month_panel:")
display(dataset_repo_month_panel.info())

print("Missing values in dataset_repo_level_busfactor:")
display(dataset_repo_level_busfactor.info())
'''
print("Missing values in dataset_vuln_discovery_rate:")
display(dataset_vuln_discovery_rate.info())
'''
print("Missing values in dataset_vuln_survival:")
display(dataset_vuln_survival.info())
'''

Missing values in dataset_vuln_discovery_rate:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 364 entries, 0 to 363
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   repo_full_name              364 non-null    object 
 1   quarter                     364 non-null    object 
 2   vuln_count                  364 non-null    int64  
 3   vuln_critical               364 non-null    int64  
 4   vuln_high                   364 non-null    int64  
 5   vuln_medium                 364 non-null    int64  
 6   vuln_low                    364 non-null    int64  
 7   vuln_severe                 364 non-null    int64  
 8   stars                       364 non-null    float64
 9   forks                       364 non-null    float64
 10  log_stars                   364 non-null    float64
 11  transparency_index          364 non-null    float64
 12  governance_index            364 non-null    f

None

'\nprint("Missing values in dataset_vuln_survival:")\ndisplay(dataset_vuln_survival.info())\n'

In [4]:
'''
print("Missing values in dataset_repo_month_panel:")
display(dataset_repo_month_panel.isnull().sum())

print("Missing values in dataset_repo_level_busfactor:")
display(dataset_repo_level_busfactor.isnull().sum())
'''
print("Missing values in dataset_vuln_discovery_rate:")
display(dataset_vuln_discovery_rate.isnull().sum())
'''
print("Missing values in dataset_vuln_survival:")
display(dataset_vuln_survival.isnull().sum())
'''

Missing values in dataset_vuln_discovery_rate:


,0
repo_full_name,0
quarter,0
vuln_count,0
vuln_critical,0
vuln_high,0
vuln_medium,0
vuln_low,0
vuln_severe,0
stars,0
forks,0


'\nprint("Missing values in dataset_vuln_survival:")\ndisplay(dataset_vuln_survival.isnull().sum())\n'

In [5]:
import statsmodels.api as sm
import numpy as np

# Drop rows where 'transparency_index' or 'vuln_count' is NaN
df_cleaned = dataset_vuln_discovery_rate.dropna(subset=['transparency_index', 'vuln_count']).copy()

# Ensure vuln_count is integer type (required for Poisson regression)
df_cleaned['vuln_count'] = df_cleaned['vuln_count'].astype(int)

print(f"Original rows: {len(dataset_vuln_discovery_rate)}")
print(f"Rows after dropping NaNs for relevant columns: {len(df_cleaned)}")
print("Missing values after cleaning:")
print(df_cleaned[['transparency_index', 'vuln_count']].isnull().sum())

Original rows: 364
Rows after dropping NaNs for relevant columns: 364
Missing values after cleaning:
transparency_index    0
vuln_count            0
dtype: int64


In [6]:
# Define the dependent variable (y) and independent variable (X)
y = df_cleaned['vuln_count']
X = df_cleaned['transparency_index']

# Add a constant to the independent variable for the intercept
X = sm.add_constant(X)

# Fit the Poisson regression model
poisson_model = sm.GLM(y, X, family=sm.families.Poisson())
poisson_results = poisson_model.fit()

# Print the summary of the regression results
print(poisson_results.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:             vuln_count   No. Observations:                  364
Model:                            GLM   Df Residuals:                      362
Model Family:                 Poisson   Df Model:                            1
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -2291.3
Date:                Sat, 07 Mar 2026   Deviance:                       3624.5
Time:                        04:02:01   Pearson chi2:                 1.33e+04
No. Iterations:                     6   Pseudo R-squ. (CS):             0.5273
Covariance Type:            nonrobust                                         
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -3.9916      0

The Poisson regression model has been fitted. Now, let's interpret the results, focusing on the `transparency_index` coefficient and its Incident Rate Ratio (IRR).

In [7]:
# Calculate Incident Rate Ratios (IRR) for each coefficient
irr = np.exp(poisson_results.params)

# Get the confidence intervals for the IRRs
conf_int = np.exp(poisson_results.conf_int())

# Create a DataFrame for better display
irr_df = pd.DataFrame({
    'IRR': irr,
    'Lower CI': conf_int[0],
    'Upper CI': conf_int[1]
})

print("Incident Rate Ratios (IRR) and 95% Confidence Intervals:")
display(irr_df)

# Extract the p-value for transparency_index
p_value_transparency = poisson_results.pvalues['transparency_index']

# Interpret the results for transparency_index
print(f"\nInterpretation for Transparency Index:")
print(f"Coefficient: {poisson_results.params['transparency_index']:.4f}")
print(f"IRR: {irr['transparency_index']:.4f}")
print(f"95% Confidence Interval for IRR: [{conf_int.loc['transparency_index', 0]:.4f}, {conf_int.loc['transparency_index', 1]:.4f}]")
print(f"P-value: {p_value_transparency:.4f}")

if p_value_transparency < 0.05:
    print("Since the p-value is less than 0.05, we reject the null hypothesis (H02a).")
    if irr['transparency_index'] > 1:
        print("The IRR for transparency_index is greater than 1, suggesting that transparency increases the rate of vulnerability discovery.")
    else:
        print("The IRR for transparency_index is less than 1, suggesting that transparency decreases the rate of vulnerability discovery.")
else:
    print("Since the p-value is greater than 0.05, we fail to reject the null hypothesis (H02a).")
    print("This suggests that transparency has no statistically significant effect on the rate of vulnerability discovery.")

Incident Rate Ratios (IRR) and 95% Confidence Intervals:


,IRR,Lower CI,Upper CI
const,0.018470,0.008609,0.039628
transparency_index,355.139359,158.319704,796.640979



Interpretation for Transparency Index:
Coefficient: 5.8725
IRR: 355.1394
95% Confidence Interval for IRR: [158.3197, 796.6410]
P-value: 0.0000
Since the p-value is less than 0.05, we reject the null hypothesis (H02a).
The IRR for transparency_index is greater than 1, suggesting that transparency increases the rate of vulnerability discovery.


The Poisson regression analysis has been completed. The `transparency_index` shows a statistically significant effect on the `vuln_count` (p-value = 0.0000). The Incident Rate Ratio (IRR) for `transparency_index` is approximately **355.14**, with a 95% confidence interval of [158.32, 796.64]. Since the p-value is less than 0.05 and the IRR is significantly greater than 1, **we reject the null hypothesis (H02a)**. This suggests that **transparency significantly increases the rate of vulnerability discovery**, supporting your alternate hypothesis (H12a) that more transparent projects surface vulnerabilities more frequently.

### Negative Binomial Regression Analysis

Now, let's perform a Negative Binomial regression to account for potential overdispersion in the `vuln_count` data.

In [8]:
# Define the dependent variable (y) and independent variable (X) - reusing from previous analysis
y_nb = df_cleaned['vuln_count']
X_nb = df_cleaned['transparency_index']

# Add a constant to the independent variable for the intercept
X_nb = sm.add_constant(X_nb)

# Fit the Negative Binomial regression model
# Note: The 'alpha' parameter in NegativeBinomial family handles overdispersion.
neg_binom_model = sm.GLM(y_nb, X_nb, family=sm.families.NegativeBinomial())
neg_binom_results = neg_binom_model.fit()

# Print the summary of the regression results
print(neg_binom_results.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:             vuln_count   No. Observations:                  364
Model:                            GLM   Df Residuals:                      362
Model Family:        NegativeBinomial   Df Model:                            1
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -906.63
Date:                Sat, 07 Mar 2026   Deviance:                       394.37
Time:                        04:02:01   Pearson chi2:                 2.13e+03
No. Iterations:                     7   Pseudo R-squ. (CS):             0.1409
Covariance Type:            nonrobust                                         
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -3.3735      0

/usr/local/lib/python3.12/dist-packages/statsmodels/genmod/families/family.py:1367: ValueWarning: Negative binomial dispersion parameter alpha not set. Using default value alpha=1.0.
  warnings.warn("Negative binomial dispersion parameter alpha not "


The Negative Binomial regression model has been fitted. Let's interpret its results, focusing on the `transparency_index` coefficient and its Incident Rate Ratio (IRR), similar to the Poisson regression.

In [9]:
# Calculate Incident Rate Ratios (IRR) for each coefficient
irr_nb = np.exp(neg_binom_results.params)

# Get the confidence intervals for the IRRs
conf_int_nb = np.exp(neg_binom_results.conf_int())

# Create a DataFrame for better display
irr_nb_df = pd.DataFrame({
    'IRR': irr_nb,
    'Lower CI': conf_int_nb[0],
    'Upper CI': conf_int_nb[1]
})

print("Incident Rate Ratios (IRR) and 95% Confidence Intervals for Negative Binomial Model:")
display(irr_nb_df)

# Extract the p-value for transparency_index
p_value_transparency_nb = neg_binom_results.pvalues['transparency_index']

# Interpret the results for transparency_index
print(f"\nInterpretation for Transparency Index (Negative Binomial Model):")
print(f"Coefficient: {neg_binom_results.params['transparency_index']:.4f}")
print(f"IRR: {irr_nb['transparency_index']:.4f}")
print(f"95% Confidence Interval for IRR: [{conf_int_nb.loc['transparency_index', 0]:.4f}, {conf_int_nb.loc['transparency_index', 1]:.4f}]")
print(f"P-value: {p_value_transparency_nb:.4f}")

if p_value_transparency_nb < 0.05:
    print("Since the p-value is less than 0.05, we reject the null hypothesis (H02a).")
    if irr_nb['transparency_index'] > 1:
        print("The IRR for transparency_index is greater than 1, suggesting that transparency increases the rate of vulnerability discovery (Negative Binomial Model).")
    else:
        print("The IRR for transparency_index is less than 1, suggesting that transparency decreases the rate of vulnerability discovery (Negative Binomial Model).")
else:
    print("Since the p-value is greater than 0.05, we fail to reject the null hypothesis (H02a).")
    print("This suggests that transparency has no statistically significant effect on the rate of vulnerability discovery (Negative Binomial Model).")

Incident Rate Ratios (IRR) and 95% Confidence Intervals for Negative Binomial Model:


,IRR,Lower CI,Upper CI
const,0.034271,0.009448,0.124319
transparency_index,182.576180,45.206613,737.371351



Interpretation for Transparency Index (Negative Binomial Model):
Coefficient: 5.2072
IRR: 182.5762
95% Confidence Interval for IRR: [45.2066, 737.3714]
P-value: 0.0000
Since the p-value is less than 0.05, we reject the null hypothesis (H02a).
The IRR for transparency_index is greater than 1, suggesting that transparency increases the rate of vulnerability discovery (Negative Binomial Model).


The model's results for `transparency_index` are statistically significant (p-value = 0.0000). The Incident Rate Ratio (IRR) for `transparency_index` in this model is approximately 182.58, with a 95% confidence interval of [45.21, 737.37]. Similar to the Poisson regression, since the p-value is less than 0.05 and the IRR is significantly greater than 1, **we reject the null hypothesis (H02a)**. This further supports the idea that **transparency increases the rate of vulnerability discovery**.




# Summary of Findings:

* **Hypothesis Tested:** Transparency increases the rate of vulnerability discovery (H12a: IRR_Transparency > 1).
* **Poisson Regression Results:** The `transparency_index` showed a highly statistically significant positive effect on `vuln_count` (p-value < 0.001). The Incident Rate Ratio (IRR) was approximately **355.14** (95% CI: [158.32, 796.64]), indicating that a higher transparency index is associated with a significantly increased rate of vulnerability discovery.
* **Negative Binomial Regression Results:** This model also found a highly statistically significant positive effect for `transparency_index` on vuln_count (p-value < 0.001). The IRR was approximately **182.58** (95% CI: [45.21, 737.37]), consistently supporting the finding that increased transparency leads to a higher rate of vulnerability discovery. This model implicitly accounted for potential overdispersion in the data.

**Conclusion:** Both models consistently lead to the rejection of the null hypothesis (H02a) and provide strong support for your alternate hypothesis (H12a): **Transparency significantly increases the rate of vulnerability discovery in projects.** This suggests that more transparent projects tend to surface vulnerabilities more frequently.